src/Npgsql.SourceGenerators/DatabaseInfoSourceGenerator.cs

```diff
@@ -0,0 +1,643 @@
+                    Oid = rangeTypeOid,
+                    ElementName = rangeSubTypeName,
+                });
+                multiRangeTypes.Add(new
+                {
+                    Name = rangeMultiTypeName,
+                    Oid = rangeMultiTypeOid,
+                    ElementName = rangeTypeName,
+                });
+            }
+        }
+
+        // There is information about 4 composite types in pg_type.dat (pg_type, pg_attribute, pg_proc, pg_class)
+        // We don't bother adding them for now since there is no pg_attribute.dat and we'd have to parse the
+        // C header file pg_attribute.h to acquire the field information we need to make them work
+        var obj = new ScriptObject();
+        obj.Import(new
+        {
+            UnreferencedBaseTypes = pgType.Where(d => !d.ContainsKey("typtype") && !d.ContainsKey("array_type_oid") && (pgRange == null || pgRange.All(r => r["rngsubtype"] != d["typname"]))).Select(t => new
+            {
+                Name = t["typname"],
+                Oid = t["oid"]
+            }).ToArray(),
+            ReferencedBaseTypes = pgType.Where(d => !d.ContainsKey("typtype") && (d.ContainsKey("array_type_oid") || (pgRange?.Any(r => r["rngsubtype"] == d["typname"]) ?? false))).Select(t => new
+            {
+                Name = t["typname"],
+                Oid = t["oid"]
+            }).ToArray(),
+            UnreferencedPseudoTypes = pgType.Where(d => d["typtype"] == "p" && !d.ContainsKey("array_type_oid") && (pgRange == null || pgRange.All(r => r["rngsubtype"] != d["typname"]))).Select(t => new
+            {
+                Name = t["typname"],
+                Oid = t["oid"]
+            }).ToArray(),
+            ReferencedPseudoTypes = pgType.Where(d => d["typtype"] == "p" && (d.ContainsKey("array_type_oid") || (pgRange?.Any(r => r["rngsubtype"] == d["typname"]) ?? false))).Select(t => new
+            {
+                Name = t["typname"],
+                Oid = t["oid"]
+            }).ToArray(),
+            RangeTypes = rangeTypes,
+            MultirangeTypes = multiRangeTypes,
+            ArrayTypes = pgType
+                .Where(d => d.ContainsKey("array_type_oid")
+                            && d["typtype"] != "c" /* Exclude arrays of composite types. See comment above. */
+                            && (pgRange != null || d["typtype"] != "r" && d["typtype"] != "m")
+                    )
+                .Select(t => new
+                {
+                    Name = "_" + t["typname"],
+                    Oid = t["array_type_oid"],
+                    ElementName = t["typname"],
+                }).ToArray(),
+        });
+        var tc = new TemplateContext(obj);
+        tc.AutoIndent = false;
+        var output = template.Render(tc);
+        context.AddSource(generatedDatabaseInfoSymbol.Name + ".Generated.cs", SourceText.From(output, Encoding.UTF8));
+    }
+
+    static List<DatFileObject>? ParseDatFile(SourceText? content)
```

Comment:
As discussed offline, I think we should consider having a tiny Perl script to fetch the PG dat files and convert them to JSON. That should be very trivial to do, would allow us to remove all parsing code here, and would also probably be less brittle if there's any new syntactical stuff in the .dat. We'd store the resulting json in source control.

Reply:
I'm still a bit reluctant.
That would add another layer of indirection, meaning we'd loose the immediate link between the source file and our generated file.

Reply:
The main arguments from my side are:

1. Cutting out a lot of new parsing code (always a good thing...)
2. The fact that the file technically isn't some format, but a programming language (Perl). So any parser we build here would be inherently brittle.

I'm not sure what we'd be losing by committing a JSON representation instead of the original .dat - in my mind we'd just have a Perl script we can run, which fetches the new version and dumps it as JSON. But maybe I'm missing something...

@vonzshik any opinion on this?

Reply:
> in my mind we'd just have a Perl script we can run, which fetches the new version and dumps it as JSON

Ok, I'll look into this - but not tonight.

Reply:
I couldn't resist.

That's the perl code.
Whether we want to trust the postgres maintainers enough to eval some file we've automatically downloaded from their git server is a matter to be debated...

```perl
#!/usr/bin/perl
use strict;
use warnings;
use LWP::Simple;
use JSON;

my $json = JSON->new->allow_nonref;
my $pg_range = get('https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob_plain;f=src/include/catalog/pg_range.dat;hb=refs/heads/REL_14_STABLE');
die "Couldn't get pg_range.dat: $!" unless defined $pg_range;
my $pg_type = get('https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob_plain;f=src/include/catalog/pg_type.dat;hb=refs/heads/REL_14_STABLE');
die "Couldn't get pg_type.dat: $!" unless defined $pg_type;

open(FH, '>', 'catalogs.json') or die $!;
print FH $json->pretty->encode( { pg_range => eval($pg_range), pg_type => eval($pg_type)} );
close(FH);
```

PS: I'm still impressed how little ceremony you need to do complex things in perl.

Reply:
`die ... unless ...` :heart: :heart: :heart:

It feels like I'm in my 20s and the world is a beautiful, endless place :rofl: 

Yeah, Perl really was (is?) great in its own way. I think it's OK to trust and eval that very specific URL - I'm assuming that's what the PG Perl scripts do as well, right?

Reply:
I'm in favor of having a CI, which runs Perl script and commits a JSON file.

Reply:
> I think it's OK to trust and eval that very specific URL - I'm assuming that's what the PG Perl scripts do as well, right?

Well, not really. They offer the feature of [preserving line information](https://github.com/postgres/postgres/blob/826be1ffb28dde083552150243dbbdcac55ab28e/src/backend/catalog/Catalog.pm#L274) from the source files so they don't eval the whole file but rather parse it in a rather hacky way and [only eval the single hash references within the array](https://github.com/postgres/postgres/blob/826be1ffb28dde083552150243dbbdcac55ab28e/src/backend/catalog/Catalog.pm#L313).
Also they don't eval a thing they downloaded from a foreign URL but rather a file that got checked out alongside their other source files so there is less of an attack vector there.

Reply:
I think manual updating of the dat is fine too...